### PR TITLE
Exponential Moving Average

### PR DESCRIPTION
**Pull request by FilippoC**
Hi,

I am working on an implementation of the Exponential Moving Average of parameters that can be used with any optimizer. It (will) work on CPU&GPU.

A previous pull-request was rejected #385. However, this feature seems to be of interest for users, see #73 and #1465. I implemented everything in the Trainer class, therefore the code is simple and located in a single place. I hope that this can be merged when finished. :-) 

TODO:
- test everything
- as of now, the EMA is computed for every parameters of the model. This introduce extra work for non-learned parameters (for example fixed embeddings). I don't know how to fix that yet, but I will look into that
- An example would be nice, probably a tagger where we can show that the EMA gets (hopefully) better results in the PTB
- update save/populate methods

### CODE REVIEW

**Comment by neubig** [comment]
Thanks! I think this looks like a better solution, as it limits the extra complexity.

**Comment by alexandres** [comment]
I think this is a good solution for EMA, but in my case (#1465) I would like a SMA as per [Polyak 92](http://www.meyn.ece.ufl.edu/archive/spm_files/Courses/ECE555-2011/555media/poljud92.pdf).

Can this be modified to accommodate that as well?

In #1465 I was thinking more of creating in-place operations on `Parameters` which would allow SMA (and EMA), being more flexible than this proposal (I believe). Not an expert on dynet so I could very well be wrong! 

You could then have a `ParameterCollection` just for the averages of `Parameters` of interest. Finally, one would need a way to copy the averages to the main `ParameterCollection` so evaluation can be run.

**Comment by neubig** [comment]
@alexandres yes, in-place operations on parameters would be a really nice thing to have. @pmichel31415 and I discussed this before and it's on our list of things to do. This is a bit of a bigger change though, so I think for the time being having these EMA trainers implemented is nice.

Currently tests are failing though, so we'll need to find the reason.

**Comment by FilippoC** [comment]
@neubig yes, I did not expect the test to run on this first commit. I'll fix everything in the days.

@alexandres Do you have the update rules somewhere ? I am a bit lazy to through the paper of Polyak. Is this the same as the averaged perceptron?

**Comment by alexandres** [comment]
@FilippoC here is the PyTorch implementation: https://github.com/pytorch/pytorch/blob/master/torch/optim/asgd.py

**Comment by FilippoC** [comment]
@alexandres Unfortunately reading code from an API that I don't know without type declaration is a little bit difficult. If you have a textual explanation with proper math, it would be easier. :-) 
Anyway, I will try to check that, but I cannot promise anything.

**Comment by alexandres** [comment]
@FilippoC No worries! The update equation from Polyak is simply:

image

and for a running sum (what we need here)

image

**Comment by pmichel31415** [comment]
Hi @FilippoC CI seems to be failing due to compilation errors for manual cuda install, do you have time to take a look?

**Comment by FilippoC** [comment]
@pmichel31415  Unfortunately I am very busy this week. I will try to finish this next week, I am sorry for the delay.
(one of my concerns at the moment with this pull request is that I was not able to have a setting where it improve results on a simple task!)

File: dynet/training.cc [hunk 1]
```diff
@@ -279,7 +370,213 @@ void Trainer::populate(std::istream& is, real lr)
...
+    extern template void Trainer::update_ema_rule_dev<Device_GPU>(const Device_GPU& dev, Tensor* ema, Tensor* p);
+    template void Trainer::update_ema_rule_dev<Device_CPU>(const Device_CPU& dev, Tensor* ema, Tensor* p);
+    void Trainer::update_ema_rule(Tensor* ema, Tensor* p)
+    {
+        if(ema->device->type == DeviceType::CPU)
+            update_ema_rule_dev(*(Device_CPU*)ema->device, ema, p);
+        else if(ema->device->type == DeviceType::GPU)
+        {
+            cudaSetDevice(((Device_GPU*) ema->device)->cuda_device_id);
+            update_rule_rule_dev(*(Device_GPU*) ema->device, ema, p);
```
Comment (alexandres) [inline dynet/training.cc]: Shouldn't this be `update_ema_rule_dev`?

File: dynet/training.cc [hunk 2]
```diff
@@ -279,7 +370,213 @@ void Trainer::populate(std::istream& is, real lr)
...
+            swap_params_to_ema_rule_dev(*(Device_GPU*)ema->device, bias_correction, save_weights, p, mem, ema);
+        }
+        else
+            throw std::runtime_error("Bad device in MyTrainer::swap_params_to_ema_rule");
+    }
+
+
+    extern template void Trainer::swap_params_to_weights_rule_dev<Device_GPU>(const Device_GPU& dev, Tensor* p, Tensor* mem);
+    template void Trainer::swap_params_to_weights_rule_dev<Device_CPU>(const Device_GPU& dev, Tensor* p, Tensor* mem);
+    void Trainer::swap_params_to_weights_rule(const Device_GPU& dev, Tensor* p, Tensor* mem)
```
Comment (alexandres) [inline dynet/training.cc]: Should be `void Trainer::swap_params_to_weights_rule(Tensor* p, Tensor* mem)`.

File: dynet/training.cc [hunk 3]
```diff
@@ -279,7 +370,213 @@ void Trainer::populate(std::istream& is, real lr)
...
+            cudaSetDevice(((Device_GPU*) ema->device)->cuda_device_id);
+            swap_params_to_ema_rule_dev(*(Device_GPU*)ema->device, bias_correction, save_weights, p, mem, ema);
+        }
+        else
+            throw std::runtime_error("Bad device in MyTrainer::swap_params_to_ema_rule");
+    }
+
+
+    extern template void Trainer::swap_params_to_weights_rule_dev<Device_GPU>(const Device_GPU& dev, Tensor* p, Tensor* mem);
+    template void Trainer::swap_params_to_weights_rule_dev<Device_CPU>(const Device_GPU& dev, Tensor* p, Tensor* mem);
```
Comment (alexandres) [inline dynet/training.cc]: Should be `const Device_CPU& dev`.

File: dynet/training.cc [hunk 4]
```diff
@@ -279,7 +370,213 @@ void Trainer::populate(std::istream& is, real lr)
...
+    void Trainer::swap_params_to_weights_rule(const Device_GPU& dev, Tensor* p, Tensor* mem)
+    {
+        if(p->device->type == DeviceType::CPU)
+            swap_params_to_weights_rule_dev(*(Device_CPU*)ema->device, p, mem);
+        else if(ema->device->type == DeviceType::GPU)
```
Comment (alexandres) [inline dynet/training.cc]: Should be `else if(p->device->type == DeviceType::GPU)`.

File: dynet/training.cc [hunk 5]
```diff
@@ -279,7 +370,213 @@ void Trainer::populate(std::istream& is, real lr)
...
+            throw std::runtime_error("Bad device in MyTrainer::swap_params_to_ema_rule");
+    }
+
+
+    extern template void Trainer::swap_params_to_weights_rule_dev<Device_GPU>(const Device_GPU& dev, Tensor* p, Tensor* mem);
+    template void Trainer::swap_params_to_weights_rule_dev<Device_CPU>(const Device_GPU& dev, Tensor* p, Tensor* mem);
+    void Trainer::swap_params_to_weights_rule(const Device_GPU& dev, Tensor* p, Tensor* mem)
+    {
+        if(p->device->type == DeviceType::CPU)
+            swap_params_to_weights_rule_dev(*(Device_CPU*)ema->device, p, mem);
```
Comment (alexandres) [inline dynet/training.cc]: Should be `swap_params_to_weights_rule_dev(*(Device_CPU*)p->device, p, mem);`

File: dynet/training.cc [hunk 6]
```diff
@@ -279,7 +370,213 @@ void Trainer::populate(std::istream& is, real lr)
...
+        Tensor& ema = ema_lp[i].all_h;
+
+        swap_params_to_ema_rule(bias_correction, save_weights, &weights, &mem, &ema);
+    }
+}
 
+void Trainer::swap_params_to_weights()
+{
+    if (!ema_params_swapped)
+        return;
```
Comment (alexandres) [inline dynet/training.cc]: @FilippoC Aside from the compilation issues above, I think the reason for why you aren't getting results is the lack of `ema_params_swapped = false;` here to undo the `ema_params_swapped = true;` on line 515. 

I made this fix and the code seems to working perfectly! :) 

Thanks a bunch for this. Using CMA as I wished was absolutely trivial. I simply changed the update rule to `tvec(*ema).device(*dev.edevice) = (real(ema_updates) * tvec(*ema) + tvec(*p)) / (real(ema_updates)+1);`.

**Comment by FilippoC** [comment]
@alexandres Thank you very much! I am currently out of work, hopefully I'll fix this monday when I will be back to work!

**Comment by neubig** [comment]
@FilippoC I see this was changed back to "WIP". Is it ready for a check? Or do things still need to be modified?

**Comment by FilippoC** [comment]
@neubig Sorry I miss your message. I want to add an example where resutls can be easily be compared. I'll try to do it this afternoon.

**Comment by FilippoC** [comment]
@neubig I made some test with the tagger, seems to work. I also updated the tagger example with command line options.

Here is the dev accuracy w.r.t. to the epoch:
moving_avg

**Comment by neubig** [comment]
Thanks! Sorry it took so long to merge.
